### ISSUE TITLE
How to create my own Transform

### ISSUE THREAD

**Issue by evakrowarska** [rusty1s/pytorch_geometric#1070]
## ❓ Questions & Help

I would like to create my own transformation for data in PyTorch Geometric, while using libraries like Scipy.
So, I had a closer look at the kNN code inside your reporsitory.
I would appreciate further explanation, such that I understand how your way of implementing transforms works, so I will be able to implement my own. 

**Code & Questions**:

```
def knn(x, y, k, batch_x=None, batch_y=None, cosine=False):

    if batch_x is None: # Question 1
        batch_x = x.new_zeros(x.size(0), dtype=torch.long)

    if batch_y is None:
        batch_y = y.new_zeros(y.size(0), dtype=torch.long)

    x = x.view(-1, 1) if x.dim() == 1 else x # Question 2
    y = y.view(-1, 1) if y.dim() == 1 else y

    assert x.dim() == 2 and batch_x.dim() == 1
    assert y.dim() == 2 and batch_y.dim() == 1
    assert x.size(1) == y.size(1)
    assert x.size(0) == batch_x.size(0)
    assert y.size(0) == batch_y.size(0)

    if x.is_cuda:
        return torch_cluster.knn_cuda.knn(x, y, k, batch_x, batch_y, cosine)

    if cosine:
        raise NotImplementedError('Cosine distance not implemented for CPU')

    # Rescale x and y.
    min_xy = min(x.min().item(), y.min().item()) # Question 3
    x, y = x - min_xy, y - min_xy

    max_xy = max(x.max().item(), y.max().item())
    x, y, = x / max_xy, y / max_xy

    # Concat batch/features to ensure no cross-links between examples exist.
    x = torch.cat([x, 2 * x.size(1) * batch_x.view(-1, 1).to(x.dtype)], dim=-1) # Question 4
    y = torch.cat([y, 2 * y.size(1) * batch_y.view(-1, 1).to(y.dtype)], dim=-1)

    tree = scipy.spatial.cKDTree(x.detach().numpy()) # Question 5
    dist, col = tree.query(y.detach().cpu(), k=k,
                           distance_upper_bound=x.size(1))
    dist = torch.from_numpy(dist).to(x.dtype)
    col = torch.from_numpy(col).to(torch.long)
    row = torch.arange(col.size(0), dtype=torch.long).view(-1, 1).repeat(1, k) # Question 6
    mask = ~torch.isinf(dist).view(-1)
    row, col = row.view(-1)[mask], col.view(-1)[mask]

    return torch.stack([row, col], dim=0)


def knn_graph(x, k, batch=None, loop=False, flow='source_to_target',
              cosine=False):
    assert flow in ['source_to_target', 'target_to_source']
    row, col = knn(x, x, k if loop else k + 1, batch, batch, cosine=cosine)
    row, col = (col, row) if flow == 'source_to_target' else (row, col)
    if not loop:
        mask = row != col
        row, col = row[mask], col[mask]
    return torch.stack([row, col], dim=0)
```

**Question 1**: Does this line mean, that if batch equals none, we have just one graph and all the features inside x belong to just one graph ? So by setting batch_x to zeroes we assign all the features to the same graph by index 0?

**Question 2:** Why is the n x 1 shape needed? I am sorry, if this question is too obvious.

**Question 3**: Rescaling x and y means that you normalize both vectors? What if both vectors are 0, then the max element is 0 as well, and we divide by 0?

**Question 4**: Can you explain these two lines, please? I see the comment above, but it is not clear to me why concatenating should resolve cross links between examples. 
Simultaneously it is not clear to me, why cross-links between examples can arise?

**Question 5**: Do you pass scipy the whole batch as an argument at once? What if we have multiple examples?
Just by looking at the code I am not able to reverse-engineer how the kNN algorithm is accomplished across the whole batch at once. 

**Question 6**: This notation is just too compact while containing too many steps for me to understand. Could you help me here, as well?

Thanks for your help in advance!

**Comment by rusty1s** [comment]
If you want to learn about transforms, it may be better to look at other examples, *e.g.*, [here](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/transforms/cartesian.py). In general, a `transform` object takes in a single graph object `data` and just returns a new one. The `knn_graph` transform is special since it also supports batches of graphs. To answer your questions:
1. Yes
2. This is just more convenient to access the `dim` of the features as `x.size(1)`.
3. That only happens in case both vectors have only zero entries, for which `knn_graph` does not make sense anyway.
4. The concatenation adds the `batch` vector to each feature. This way, `knn` can be processed in a batch, while the additional dimension ensures that `knn` will not reference any node outside the batch.
5. Yes, see answer 4.
6. I am not sure to which code part you are referring to.

**Comment by evakrowarska** [comment]
Thank you for this detailed answer! It helped me a lot.
For question 6, I meant the code line, where the comment of Question 6 is in:   
`row = torch.arange(col.size(0), dtype=torch.long).view(-1, 1).repeat(1, k) `

**Comment by rusty1s** [comment]
That just create the `row` tensor, i.e., `[0, 0, 0, 0, 1, 1, 1, 1, ...]` for `k = 4`.

**Comment by evakrowarska** [comment]
Alright, thanks!
